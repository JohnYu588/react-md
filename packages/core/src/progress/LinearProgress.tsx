import { cnb } from "cnbuilder";
import type { CSSProperties, HTMLAttributes } from "react";
import { forwardRef, useMemo } from "react";
import { useEnsuredId } from "../useEnsuredId";
import { bem, getPercentage } from "../utils";
import type { ProgressProps } from "./types";

export interface LinearProgressProps
  extends Omit<HTMLAttributes<HTMLDivElement>, "id">,
    ProgressProps {
  /**
   * An optional style to apply to the progress bar. This will be merged with
   * the current width or height tracking the progress when a `value` is also
   * provided.
   */
  barStyle?: CSSProperties;

  /**
   * An optional className to apply to the progress bar.
   */
  barClassName?: string;

  /**
   * Boolean if the progress should be reversed. This will change the progress
   * direction from `left-to-right` to be `right-to-left`. If the current
   * language is a rtl language and this prop is enabled, the direction will
   * still be `left-to-right`.
   *
   * @defaultValue `false`
   */
  reverse?: boolean;

  /**
   * Boolean if the progress should be vertical instead of horizontal.  When
   * this prop is set, you should also set the `verticalHeight` prop to a height
   * value you want for your progress bar.
   *
   * @defaultValue `false`
   */
  vertical?: boolean;

  /**
   * Since there isn't really a good way to have "auto height", you'll need to
   * manually set the progress bar's height with this prop to some pixel value.
   * If you'd prefer to set the height in Sass/css, set this value to `null`
   * instead since this value would be passed down as a `height` inline style.
   *
   * @defaultValue `240`
   */
  verticalHeight?: number | null;
}

const styles = bem("rmd-linear-progress");

export const LinearProgress = forwardRef<HTMLSpanElement, LinearProgressProps>(
  function LinearProgress(props, ref) {
    const {
      id: propId,
      style: propStyle,
      className,
      barStyle: propBarStyle,
      barClassName,
      min = 0,
      max = 100,
      value,
      reverse = false,
      disableTransition = false,
      vertical = false,
      verticalHeight = 240,
      ...remaining
    } = props;

    const id = useEnsuredId(propId, "linear-progress");
    const style = useMemo(() => {
      if (!vertical || verticalHeight === null) {
        return propStyle;
      }

      return {
        ...propStyle,
        height: verticalHeight,
      };
    }, [propStyle, vertical, verticalHeight]);

    let progress: number | undefined;
    if (typeof value === "number") {
      progress = getPercentage({ min, max, value, validate: true });
    }
    const barStyle = useMemo(() => {
      if (typeof progress !== "number") {
        return propBarStyle;
      }

      const key = vertical ? "height" : "width";
      return {
        ...propBarStyle,
        [key]: `${progress * 100}%`,
      };
    }, [progress, propBarStyle, vertical]);

    const determinate = typeof progress === "number";
    const indeterminate = !determinate;
    return (
      <span
        {...remaining}
        id={id}
        ref={ref}
        style={style}
        role="progressbar"
        aria-valuemin={min}
        aria-valuemax={max}
        aria-valuenow={value}
        className={cnb(
          styles({
            vertical,
            horizontal: !vertical,
            determinate,
            indeterminate,
          }),
          className
        )}
      >
        <span
          style={barStyle}
          className={cnb(
            styles("bar", {
              vertical,
              "vertical-reverse": vertical && reverse,
              horizontal: !vertical,
              "horizontal-reverse": !vertical && reverse,
              animate: !disableTransition && determinate,
              determinate,
              indeterminate,
              "determinate-reverse": determinate && reverse && !vertical,
              "determinate-vertical-reverse":
                determinate && reverse && vertical,
              "indeterminate-reverse": indeterminate && reverse && !vertical,
              "indeterminate-vertical": indeterminate && vertical,
              "indeterminate-vertical-reverse":
                indeterminate && reverse && vertical,
            }),
            barClassName
          )}
        />
      </span>
    );
  }
);
