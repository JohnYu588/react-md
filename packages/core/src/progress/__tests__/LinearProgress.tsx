import { describe, expect, it } from "@jest/globals";
import { createRef } from "react";
import { render, screen } from "../../test-utils/index.js";

import { LinearProgress } from "../LinearProgress.js";

describe("LinearProgress", () => {
  it("should apply the correct styling, HTMLAttributes, and allow a ref", () => {
    const ref = createRef<HTMLSpanElement>();
    const props = {
      ref,
    } as const;

    const { rerender } = render(<LinearProgress {...props} />);

    const element = screen.getByRole("progressbar");

    expect(ref.current).toBeInstanceOf(HTMLSpanElement);
    expect(ref.current).toBe(element);
    expect(element).toMatchSnapshot();

    rerender(
      <LinearProgress
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(element).toMatchSnapshot();

    rerender(
      <LinearProgress
        {...props}
        barStyle={{ color: "orange" }}
        barClassName="custom-bar-class-name"
      />
    );
    expect(element).toMatchSnapshot();
  });

  it("should allow the id to be overridden", () => {
    render(<LinearProgress id="custom-id" />);
    const progress = screen.getByRole("progressbar");
    expect(progress).toHaveAttribute("id", "custom-id");
  });

  it("should be able to render vertically", () => {
    const { rerender } = render(<LinearProgress vertical />);
    const progress = screen.getByRole("progressbar");
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress vertical verticalHeight={null} />);
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress vertical verticalHeight={500} />);
    expect(progress).toMatchSnapshot();
  });

  it("should be able to reverse the animation by enabling the reverse prop", () => {
    const { rerender } = render(<LinearProgress reverse />);
    const progress = screen.getByRole("progressbar");
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress reverse vertical />);
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress reverse value={50} />);
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress reverse value={50} vertical />);
    expect(progress).toMatchSnapshot();
  });

  it("should render as a determinate progress bar when the value is provided", () => {
    const { rerender } = render(<LinearProgress value={30} />);
    const progress = screen.getByRole("progressbar");
    expect(progress).toHaveAttribute("aria-valuenow", "30");
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress value={50} />);
    expect(progress).toHaveAttribute("aria-valuenow", "50");
    expect(progress).toMatchSnapshot();
  });

  it("should allow the transition to be disabled for determinate progress bars", () => {
    const { rerender } = render(
      <LinearProgress value={30} disableTransition />
    );
    const progress = screen.getByRole("progressbar");
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress disableTransition />);
    expect(progress).toMatchSnapshot();
  });

  it("should be able to render as any of the theme colors", () => {
    const { rerender } = render(<LinearProgress theme="primary" />);
    const progress = screen.getByRole("progressbar");
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress theme="secondary" />);
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress theme="warning" />);
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress theme="success" />);
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress theme="error" />);
    expect(progress).toMatchSnapshot();

    rerender(<LinearProgress theme="current-color" />);
    expect(progress).toMatchSnapshot();
  });
});
