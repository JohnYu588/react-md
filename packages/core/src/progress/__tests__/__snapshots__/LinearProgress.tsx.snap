// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinearProgress should allow the transition to be disabled for determinate progress bars 1`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="30"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--determinate rmd-primary-color"
  id="linear-progress-:r5:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--determinate"
    style="width: 30%;"
  />
</span>
`;

exports[`LinearProgress should allow the transition to be disabled for determinate progress bars 2`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate rmd-primary-color"
  id="linear-progress-:r5:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--indeterminate"
    style=""
  />
</span>
`;

exports[`LinearProgress should apply the correct styling, HTMLAttributes, and allow a ref 1`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate rmd-primary-color"
  id="linear-progress-:r0:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--indeterminate"
  />
</span>
`;

exports[`LinearProgress should apply the correct styling, HTMLAttributes, and allow a ref 2`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate rmd-primary-color custom-class-name"
  id="linear-progress-:r0:"
  role="progressbar"
  style="color: white;"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--indeterminate"
  />
</span>
`;

exports[`LinearProgress should apply the correct styling, HTMLAttributes, and allow a ref 3`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate rmd-primary-color"
  id="linear-progress-:r0:"
  role="progressbar"
  style=""
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--indeterminate custom-bar-class-name"
    style="color: orange;"
  />
</span>
`;

exports[`LinearProgress should be able to render as any of the theme colors 1`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate rmd-primary-color"
  id="linear-progress-:r6:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--indeterminate"
  />
</span>
`;

exports[`LinearProgress should be able to render as any of the theme colors 2`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate rmd-secondary-color"
  id="linear-progress-:r6:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--indeterminate"
  />
</span>
`;

exports[`LinearProgress should be able to render as any of the theme colors 3`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate rmd-warning-color"
  id="linear-progress-:r6:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--indeterminate"
  />
</span>
`;

exports[`LinearProgress should be able to render as any of the theme colors 4`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate rmd-success-color"
  id="linear-progress-:r6:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--indeterminate"
  />
</span>
`;

exports[`LinearProgress should be able to render as any of the theme colors 5`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate rmd-error-color"
  id="linear-progress-:r6:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--indeterminate"
  />
</span>
`;

exports[`LinearProgress should be able to render as any of the theme colors 6`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate"
  id="linear-progress-:r6:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--indeterminate"
  />
</span>
`;

exports[`LinearProgress should be able to render vertically 1`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--vertical rmd-linear-progress--indeterminate rmd-primary-color"
  id="linear-progress-:r2:"
  role="progressbar"
  style="height: 240px;"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--vertical rmd-linear-progress__bar--indeterminate rmd-linear-progress__bar--indeterminate-vertical"
  />
</span>
`;

exports[`LinearProgress should be able to render vertically 2`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--vertical rmd-linear-progress--indeterminate rmd-primary-color"
  id="linear-progress-:r2:"
  role="progressbar"
  style=""
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--vertical rmd-linear-progress__bar--indeterminate rmd-linear-progress__bar--indeterminate-vertical"
  />
</span>
`;

exports[`LinearProgress should be able to render vertically 3`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--vertical rmd-linear-progress--indeterminate rmd-primary-color"
  id="linear-progress-:r2:"
  role="progressbar"
  style="height: 500px;"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--vertical rmd-linear-progress__bar--indeterminate rmd-linear-progress__bar--indeterminate-vertical"
  />
</span>
`;

exports[`LinearProgress should be able to reverse the animation by enabling the reverse prop 1`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--indeterminate rmd-primary-color"
  id="linear-progress-:r3:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--horizontal-reverse rmd-linear-progress__bar--indeterminate rmd-linear-progress__bar--indeterminate-reverse"
  />
</span>
`;

exports[`LinearProgress should be able to reverse the animation by enabling the reverse prop 2`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-linear-progress rmd-linear-progress--vertical rmd-linear-progress--indeterminate rmd-primary-color"
  id="linear-progress-:r3:"
  role="progressbar"
  style="height: 240px;"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--vertical rmd-linear-progress__bar--vertical-reverse rmd-linear-progress__bar--indeterminate rmd-linear-progress__bar--indeterminate-vertical rmd-linear-progress__bar--indeterminate-vertical-reverse"
  />
</span>
`;

exports[`LinearProgress should be able to reverse the animation by enabling the reverse prop 3`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="50"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--determinate rmd-primary-color"
  id="linear-progress-:r3:"
  role="progressbar"
  style=""
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--horizontal-reverse rmd-linear-progress__bar--animate rmd-linear-progress__bar--determinate rmd-linear-progress__bar--determinate-reverse"
    style="width: 50%;"
  />
</span>
`;

exports[`LinearProgress should be able to reverse the animation by enabling the reverse prop 4`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="50"
  class="rmd-linear-progress rmd-linear-progress--vertical rmd-linear-progress--determinate rmd-primary-color"
  id="linear-progress-:r3:"
  role="progressbar"
  style="height: 240px;"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--vertical rmd-linear-progress__bar--vertical-reverse rmd-linear-progress__bar--animate rmd-linear-progress__bar--determinate rmd-linear-progress__bar--determinate-vertical-reverse"
    style="height: 50%;"
  />
</span>
`;

exports[`LinearProgress should render as a determinate progress bar when the value is provided 1`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="30"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--determinate rmd-primary-color"
  id="linear-progress-:r4:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--animate rmd-linear-progress__bar--determinate"
    style="width: 30%;"
  />
</span>
`;

exports[`LinearProgress should render as a determinate progress bar when the value is provided 2`] = `
<span
  aria-valuemax="100"
  aria-valuemin="0"
  aria-valuenow="50"
  class="rmd-linear-progress rmd-linear-progress--horizontal rmd-linear-progress--determinate rmd-primary-color"
  id="linear-progress-:r4:"
  role="progressbar"
>
  <span
    class="rmd-linear-progress__bar rmd-linear-progress__bar--horizontal rmd-linear-progress__bar--animate rmd-linear-progress__bar--determinate"
    style="width: 50%;"
  />
</span>
`;
